### PR TITLE
BL-2678 restrict book cmds

### PR DESCRIPTION
**Pull request by gmartin7**
Eliminates context menu for books distributed with Bloom. Allows deletion of books downloaded from BloomLibrary.org. Limits context menu in ways set out in BL-2678.

Also disallows using 'templates' or some case variant as a new collection name.

### CODE REVIEW

File: src/BloomExe/Book/Book.cs [hunk 1]
```diff
@@ -581,10 +581,16 @@ private HtmlDom GetErrorDom()
 			return new HtmlDom(builder.ToString());
 		}
 
+		private bool IsDownloaded
+		{
+			get { return FolderPath.Contains(BookCollection.DownloadedBooksCollectionNameInEnglish); }
```
Comment (JohnThomson) [inline src/BloomExe/Book/Book.cs]: Why not just use BookTransfer.DownloadFolder and test whether it is for sure the folder we use for Downloads? This COULD fail if the user made one with that name..

File: src/BloomExe/Collection/BookCollection.cs [hunk 2]
```diff
@@ -189,7 +187,34 @@ protected Color CoverColor
 			set { throw new NotImplementedException(); }
 		}
 
-		public static string DownloadedBooksCollectionNameInEnglish = "Books From BloomLibrary.org";
+		public const string DownloadedBooksCollectionNameInEnglish = "Books From BloomLibrary.org";
+		private const string InstalledTemplatesCollectionNameInEnglish = "Templates";
+
+		public bool IsDownloaded { get { return Name == DownloadedBooksCollectionNameInEnglish; } }
```
Comment (JohnThomson) [inline src/BloomExe/Collection/BookCollection.cs]: The collection isn't downloaded...it's the collection reserved to contain things that are downloaded.

File: src/BloomExe/Collection/BookCollection.cs [hunk 3]
```diff
@@ -189,7 +187,34 @@ protected Color CoverColor
...
+
+		private string FactoryDir
```
Comment (JohnThomson) [inline src/BloomExe/Collection/BookCollection.cs]: Is this trying to find something different from ProjectContext.FactoryCollectionsDirectory or ProjectContext.GetInstalledCollectionsDirectory()? It's confusing because I think Basic Book is normally in the former, but this code seems to be looking for it in the latter. But then, maybe I'm the one confused...these locations are different for dev and end-user machines, and I seem to have gotten it wrong in one recent PR. Still, I think we should try to have only one bit of code that knows where these directories are, and I think ProjectContext is a pretty good place for it if it isn't already there.

File: src/BloomExe/CollectionCreating/CollectionNameControl.cs [hunk 4]
```diff
@@ -75,9 +75,7 @@ private bool GetNameIsOk()
 				}
 
 				_collectionInfo.PathToSettingsFile = CollectionSettings.GetPathForNewSettings(_destinationDirectory, _collectionNameControl.Text);
-				if (DestinationAlreadyExists)
-					return false;
-				return true;
+				return !DestinationAlreadyExists && _collectionNameControl.Text.ToLowerInvariant() != "templates";
```
Comment (JohnThomson) [inline src/BloomExe/CollectionCreating/CollectionNameControl.cs]: @hatton, should we give the user some explanation of why his button is disabled? Not that we're doing it if he types the name of an existing collection, I guess.

File: src/BloomExe/CollectionTab/LibraryListView.cs [hunk 5]
```diff
@@ -92,19 +93,39 @@ enum ButtonManagementStage
...
+				if (menuItem == _openFolderOnDisk)
+					continue; // leave this one on (for both BloomLibrary and BloomPack books)
+				menuItem.Visible = false;
+			}
 		}
 
 		void _bookContextMenu_Closed(object sender, ToolStripDropDownClosedEventArgs e)
 		{
-			// make these controls visible again so they are available in the right-click menu
-			toolStripSeparator1.Visible = true;
```
Comment (JohnThomson) [inline src/BloomExe/CollectionTab/LibraryListView.cs]: I suppose we don't need to hide any separators to make it look right in any of the variations?

File: src/BloomExe/CollectionTab/LibraryListView.cs [hunk 6]
```diff
@@ -759,13 +782,6 @@ private void SelectBook(BookInfo bookInfo)
 
 		private Book.Book SelectedBook
 		{
-			set
```
Comment (JohnThomson) [inline src/BloomExe/CollectionTab/LibraryListView.cs]: Doesn't seem related to task...maybe you just found this to be redundant when dealing with everything that used the Tag?

File: src/BloomExe/CollectionTab/LibraryListView.cs [hunk 7]
```diff
@@ -856,7 +872,7 @@ bool IsUsableBook(Button bookButton)
 			// Eithe way, the basic idea is that books in the main collection you are now editing are always usable.
 			if (bookButton.Parent == _primaryCollectionFlow)
 				return true;
-			var bookInfo = (BookInfo) bookButton.Tag;
+			var bookInfo = (bookButton.Tag as BookButtonInfo).BookInfo;
```
Comment (JohnThomson) [inline src/BloomExe/CollectionTab/LibraryListView.cs]: Getting very repetitive...it's probably not worth it now, but I would have been inclined, while changing all these places, to encapsulate in a method the notion of how to get a BookInfo from the tag of a button.

File: src/BloomExe/CollectionTab/LibraryListView.cs [hunk 8]
```diff
@@ -1210,4 +1231,27 @@ private static void CopyDirectory(string sourceDir, string targetDir)
...
+
+	internal class BookButtonInfo
+	{
+		private readonly BookInfo _bookInfo;
+		internal BookInfo BookInfo { get { return _bookInfo; } }
+
+		private readonly BookCollection _collection;
+
+		// isVernacular meaning part of the current TheOneEditableCollection
+		private readonly bool _isVernacularBook;
```
Comment (JohnThomson) [inline src/BloomExe/CollectionTab/LibraryListView.cs]: Don't like the name much. If someone is authoring a template in English, I would not think of its as a 'vernacular' book. IsIn{TheOne}EditableCollection?

File: src/BloomExe/CollectionTab/LibraryListView.cs (hunk 6 above)
Comment (gmartin7) [inline src/BloomExe/CollectionTab/LibraryListView.cs]: R# said the setter was unused and I couldn't find any usages either.

File: src/BloomExe/Book/Book.cs (hunk 1 above)
Comment (gmartin7) [inline src/BloomExe/Book/Book.cs]: done

File: src/BloomExe/Collection/BookCollection.cs (hunk 2 above)
Comment (gmartin7) [inline src/BloomExe/Collection/BookCollection.cs]: changed to ContainsDownloadedBooks

File: src/BloomExe/Collection/BookCollection.cs (hunk 3 above)
Comment (gmartin7) [inline src/BloomExe/Collection/BookCollection.cs]: This simplified things greatly. I had to make ProjectContext.FactoryCollectionsDirectory public, but I think it's worth it. GetInstalledCollectionsDirectory() is finding the place where bloompacks get put, which is different than factory installed stuff.

File: src/BloomExe/CollectionTab/LibraryListView.cs (hunk 8 above)
Comment (gmartin7) [inline src/BloomExe/CollectionTab/LibraryListView.cs]: Ha! Good catch. It's even easier than that! _bookInfo already has an IsEditable property (with the same definition). So now BookButtonInfo.IsEditable returns _bookInfo.IsEditable.

File: src/BloomExe/CollectionTab/LibraryListView.cs [hunk 9]
```diff
@@ -92,19 +91,48 @@ enum ButtonManagementStage
...
 
 			SetupBookDropdownIcon();
 			_bookContextMenu.Closed += _bookContextMenu_Closed;
+			_bookContextMenu.Opening += _bookContextMenu_Opening;
+		}
+
+		// adjust the context menu item visibility based on what sort of collection we're in
+		void _bookContextMenu_Opening(object sender, CancelEventArgs e)
+		{
+			var btn = (sender as ContextMenuStrip).SourceControl as Button;
```
Comment (JohnThomson) [inline src/BloomExe/CollectionTab/LibraryListView.cs]: Where are the menu items created? Seems it would be useful to have a cross-reference comment there saying that currently these menu items are all disabled for factory books, and most for bloom library...if someone adds a new one they should be prompted to think about which collections should have it.

Comment (gmartin7) [inline src/BloomExe/CollectionTab/LibraryListView.cs]: The menu items are mostly created in the designer... so not really a good place to put extra comments, as they might get overwritten later. :(

File: src/BloomExe/CollectionTab/LibraryListView.cs [hunk 10]
```diff
@@ -702,9 +730,11 @@ private void HighlightBookButtonAndShowContextMenuButton(BookInfo bookInfo)
 		{
 			foreach (var btn in AllBookButtons())
 			{
-				if (btn.Tag == bookInfo)
+				var bookButtonInfo = btn.Tag as BookButtonInfo;
+				if (bookButtonInfo.BookInfo == bookInfo)
 				{
-					btn.Paint += btn_Paint;
+					// BL-2678 don't display menu triangle if there's no menu to display
+					if(!bookButtonInfo.IsFactoryTemplate) btn.Paint += btn_Paint;
```
Comment (JohnThomson) [inline src/BloomExe/CollectionTab/LibraryListView.cs]: Hmmm...now we have yet another place that 'knows' that factory templates have no menu items. Would it be worth adding a method to BookButtonInfo, something like HasNoTriangleMenuItems, so the knowledge that factory templates have this property is in just one place?

**Comment by JohnThomson** [comment]
Looks much better...couple of minor points (sorry I missed before).

File: src/BloomExe/CollectionTab/LibraryListView.cs (hunk 5 above)
Comment (gmartin7) [inline src/BloomExe/CollectionTab/LibraryListView.cs]: nope

File: src/BloomExe/CollectionTab/LibraryListView.cs (hunk 10 above)
Comment (gmartin7) [inline src/BloomExe/CollectionTab/LibraryListView.cs]: good point. I decided to just call it HasNoContextMenu.
